crates/ruff_linter/src/rules/flake8_pyi/rules/custom_type_var_return_type.rs

```diff
@@ -230,3 +236,183 @@ fn is_likely_private_typevar(type_var_name: &str, type_params: Option<&TypeParam
         })
     })
 }
+
+fn add_diagnostic(
+    checker: &mut Checker,
+    method_name: String,
+    type_params: Option<&TypeParams>,
+    parameters: &Parameters,
+    returns: &Expr,
+) {
+    let in_stub = checker.source_type.is_stub();
+
+    let mut diagnostic = Diagnostic::new(
+        CustomTypeVarReturnType {
+            method_name,
+            in_stub,
+        },
+        returns.range(),
+    );
+
+    // See `replace_custom_typevar_with_self`'s doc comment
+    if in_stub {
+        let fix = replace_custom_typevar_with_self(checker, type_params, parameters, returns);
+        diagnostic.set_fix(fix);
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+/// Add a "Replace with `Self`" fix that does the following:
+///
+/// * Import `Self` if necessary
+/// * Remove the first parameter's annotation
+/// * Replace the return annotation with `Self`
+/// * Replace other uses of the original type variable elsewhere in the signature with `Self`
+/// * Remove that type variable from the PEP 695 type parameter list
+///
+/// This fix cannot be suggested for non-stubs,
+/// as a non-stub fix would have to deal with references in body/at runtime as well,
+/// which is substantially harder and requires a type-aware backend.
+///
+/// The fourth step above has the same problem.
+/// This function thus only does replacements for the simplest of cases
+/// and will mark the fix as unsafe if an annotation cannot be handled.
+fn replace_custom_typevar_with_self(
+    checker: &Checker,
+    type_params: Option<&TypeParams>,
+    parameters: &Parameters,
+    returns: &Expr,
+) -> Fix {
+    let typevar_name = returns.as_name_expr().unwrap().id();
+    let mut fix_applicability = Applicability::Safe;
+
+    let mut all_edits = vec![
+        replace_return_annotation_with_self(returns),
+        remove_first_parameter_annotation(parameters),
+    ];
+
+    if let Some(edit) = import_self(checker, returns.range()) {
+        all_edits.push(edit);
+    }
+
+    if let Some(edit) = remove_typevar_declaration(type_params, typevar_name) {
+        all_edits.push(edit);
+    }
+
+    match replace_typevar_usages_with_self(parameters, typevar_name) {
+        Ok(mut edits) => all_edits.append(&mut edits),
+        Err(mut edits) => {
+            fix_applicability = Applicability::Unsafe;
+            all_edits.append(&mut edits)
+        }
+    }
+
+    let (first, rest) = (all_edits.remove(0), all_edits);
```

Comment:
To avoid relocating the entire vec (although small)

```suggestion
    let (first, rest) = (all_edits.swap_remove(0), all_edits);
```